// @oai/sky/dist/project/cua/sky_js/src/targets/windows/internal/helper_transport.js
import { execFile } from "node:child_process";
import { appendFile, mkdir } from "node:fs/promises";
import { dirname, join } from "node:path";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);
const logPath = join(
  process.env.LOCALAPPDATA || process.env.TEMP || ".",
  "OpenAI",
  "Codex",
  "computer-use-local-helper.log",
);

async function log(entry) {
  try {
    await mkdir(dirname(logPath), { recursive: true });
    await appendFile(logPath, `${new Date().toISOString()} ${JSON.stringify(entry)}\n`, "utf8");
  } catch {
    // Logging must never break Computer Use requests.
  }
}

function encodePowerShell(script) {
  return Buffer.from(script, "utf16le").toString("base64");
}

async function runPowerShell(script, timeout = 30000) {
  const { stdout } = await execFileAsync(
    "powershell.exe",
    ["-NoProfile", "-NonInteractive", "-ExecutionPolicy", "Bypass", "-EncodedCommand", encodePowerShell(script)],
    {
      encoding: "utf8",
      env: process.env,
      timeout,
      windowsHide: true,
      maxBuffer: 64 * 1024 * 1024,
    },
  );
  const text = stdout.trim();
  return text.length === 0 ? null : JSON.parse(text);
}

function numberFrom(params, names, fallback = 0) {
  for (const name of names) {
    const value = params?.[name];
    if (typeof value === "number" && Number.isFinite(value)) return value;
    if (typeof value === "string" && value.trim() !== "" && Number.isFinite(Number(value))) return Number(value);
  }
  return fallback;
}

function buttonFrom(params) {
  const raw = String(params?.button || params?.mouseButton || "left").toLowerCase();
  if (raw.includes("right")) return "right";
  if (raw.includes("middle")) return "middle";
  return "left";
}

function keyFrom(params) {
  return String(params?.key || params?.keys || params?.text || params?.value || "");
}

function textFrom(params) {
  return String(params?.text ?? params?.value ?? params?.input ?? "");
}

const user32Script = `
Add-Type -TypeDefinition @"
using System;
using System.Runtime.InteropServices;
public static class CodexUser32 {
  [DllImport("user32.dll")] public static extern bool SetCursorPos(int X, int Y);
  [DllImport("user32.dll")] public static extern void mouse_event(uint dwFlags, uint dx, uint dy, int dwData, UIntPtr dwExtraInfo);
}
"@
`;

function mouseFlags(button, action) {
  if (button === "right") return action === "down" ? "0x0008" : "0x0010";
  if (button === "middle") return action === "down" ? "0x0020" : "0x0040";
  return action === "down" ? "0x0002" : "0x0004";
}

async function screenshot() {
  return await runPowerShell(`
Add-Type -AssemblyName System.Windows.Forms
Add-Type -AssemblyName System.Drawing
$bounds = [System.Windows.Forms.SystemInformation]::VirtualScreen
$bitmap = New-Object System.Drawing.Bitmap $bounds.Width, $bounds.Height
$graphics = [System.Drawing.Graphics]::FromImage($bitmap)
$graphics.CopyFromScreen($bounds.Left, $bounds.Top, 0, 0, $bounds.Size)
$stream = New-Object System.IO.MemoryStream
$bitmap.Save($stream, [System.Drawing.Imaging.ImageFormat]::Png)
$graphics.Dispose()
$bitmap.Dispose()
$bytes = $stream.ToArray()
$stream.Dispose()
[Console]::OutputEncoding = [System.Text.Encoding]::UTF8
[Console]::Write((ConvertTo-Json -Compress @{
  mimeType = "image/png"
  data = [Convert]::ToBase64String($bytes)
  width = $bounds.Width
  height = $bounds.Height
  left = $bounds.Left
  top = $bounds.Top
}))
`, 30000);
}

async function screenInfo() {
  return await runPowerShell(`
Add-Type -AssemblyName System.Windows.Forms
$bounds = [System.Windows.Forms.SystemInformation]::VirtualScreen
[Console]::OutputEncoding = [System.Text.Encoding]::UTF8
[Console]::Write((ConvertTo-Json -Compress @{
  width = $bounds.Width
  height = $bounds.Height
  left = $bounds.Left
  top = $bounds.Top
}))
`);
}

async function moveMouse(params) {
  const x = Math.round(numberFrom(params, ["x", "X", "left"]));
  const y = Math.round(numberFrom(params, ["y", "Y", "top"]));
  return await runPowerShell(`
${user32Script}
[CodexUser32]::SetCursorPos(${x}, ${y}) | Out-Null
[Console]::Write('{"ok":true}')
`);
}

async function clickMouse(params, count = 1) {
  const x = Math.round(numberFrom(params, ["x", "X", "left"], Number.NaN));
  const y = Math.round(numberFrom(params, ["y", "Y", "top"], Number.NaN));
  const button = buttonFrom(params);
  const down = mouseFlags(button, "down");
  const up = mouseFlags(button, "up");
  const maybeMove = Number.isFinite(x) && Number.isFinite(y) ? `[CodexUser32]::SetCursorPos(${x}, ${y}) | Out-Null` : "";
  return await runPowerShell(`
${user32Script}
${maybeMove}
for ($i = 0; $i -lt ${count}; $i++) {
  [CodexUser32]::mouse_event(${down}, 0, 0, 0, [UIntPtr]::Zero)
  Start-Sleep -Milliseconds 35
  [CodexUser32]::mouse_event(${up}, 0, 0, 0, [UIntPtr]::Zero)
  Start-Sleep -Milliseconds 70
}
[Console]::Write('{"ok":true}')
`);
}

async function dragMouse(params) {
  const fromX = Math.round(numberFrom(params, ["fromX", "startX", "x1", "x"]));
  const fromY = Math.round(numberFrom(params, ["fromY", "startY", "y1", "y"]));
  const toX = Math.round(numberFrom(params, ["toX", "endX", "x2"]));
  const toY = Math.round(numberFrom(params, ["toY", "endY", "y2"]));
  return await runPowerShell(`
${user32Script}
[CodexUser32]::SetCursorPos(${fromX}, ${fromY}) | Out-Null
Start-Sleep -Milliseconds 80
[CodexUser32]::mouse_event(0x0002, 0, 0, 0, [UIntPtr]::Zero)
Start-Sleep -Milliseconds 120
[CodexUser32]::SetCursorPos(${toX}, ${toY}) | Out-Null
Start-Sleep -Milliseconds 120
[CodexUser32]::mouse_event(0x0004, 0, 0, 0, [UIntPtr]::Zero)
[Console]::Write('{"ok":true}')
`);
}

async function scrollMouse(params) {
  const delta = Math.round(numberFrom(params, ["delta", "wheelDelta"], 0) || -120 * numberFrom(params, ["amount", "clicks"], 1));
  return await runPowerShell(`
${user32Script}
[CodexUser32]::mouse_event(0x0800, 0, 0, ${delta}, [UIntPtr]::Zero)
[Console]::Write('{"ok":true}')
`);
}

function sendKeysLiteral(text) {
  return text
    .replaceAll("{", "{{}")
    .replaceAll("}", "{}}")
    .replaceAll("+", "{+}")
    .replaceAll("^", "{^}")
    .replaceAll("%", "{%}")
    .replaceAll("~", "{~}")
    .replaceAll("(", "{(}")
    .replaceAll(")", "{)}")
    .replaceAll("[", "{[}")
    .replaceAll("]", "{]}")
    .replaceAll("\n", "{ENTER}");
}

function normalizeKey(key) {
  const value = String(key).trim();
  const upper = value.toUpperCase();
  const aliases = {
    ENTER: "{ENTER}",
    RETURN: "{ENTER}",
    ESC: "{ESC}",
    ESCAPE: "{ESC}",
    TAB: "{TAB}",
    BACKSPACE: "{BACKSPACE}",
    DELETE: "{DELETE}",
    DEL: "{DELETE}",
    SPACE: " ",
    UP: "{UP}",
    DOWN: "{DOWN}",
    LEFT: "{LEFT}",
    RIGHT: "{RIGHT}",
    HOME: "{HOME}",
    END: "{END}",
    PAGEUP: "{PGUP}",
    PAGEDOWN: "{PGDN}",
  };
  if (aliases[upper]) return aliases[upper];
  if (/^F([1-9]|1[0-2])$/.test(upper)) return `{${upper}}`;
  return sendKeysLiteral(value);
}

async function sendKeys(keys) {
  const encoded = Buffer.from(keys, "utf8").toString("base64");
  return await runPowerShell(`
Add-Type -AssemblyName System.Windows.Forms
$keys = [System.Text.Encoding]::UTF8.GetString([Convert]::FromBase64String("${encoded}"))
[System.Windows.Forms.SendKeys]::SendWait($keys)
[Console]::Write('{"ok":true}')
`);
}

async function typeText(params) {
  return await sendKeys(sendKeysLiteral(textFrom(params)));
}

async function keypress(params) {
  return await sendKeys(normalizeKey(keyFrom(params)));
}

export class WindowsHelperTransport {
  constructor({ helperArgs = [], helperCommand = null } = {}) {
    this.helperArgs = helperArgs;
    this.helperCommand = helperCommand;
    log({ event: "transport-created", helperCommand, helperArgs }).catch(() => {});
  }

  async request(method, params = {}, options = {}) {
    await log({ event: "request", method, params, hasTurnMetadata: !!options?.codexTurnMetadata });
    const name = String(method || "").replace(/[-_]/g, "").toLowerCase();
    if (name === "ping") return "pong";
    if (["screenshot", "takescreenshot", "capture", "captureimage", "capturescreen", "screencapture"].includes(name)) return await screenshot(params);
    if (["screeninfo", "getscreeninfo", "displays", "getdisplays", "screenstate"].includes(name)) return await screenInfo(params);
    if (["movemouse", "mousemove", "move"].includes(name)) return await moveMouse(params);
    if (["click", "mouseclick", "clickmouse"].includes(name)) return await clickMouse(params, 1);
    if (["doubleclick", "mousedoubleclick"].includes(name)) return await clickMouse(params, 2);
    if (["drag", "mousedrag", "dragmouse"].includes(name)) return await dragMouse(params);
    if (["scroll", "mousescroll", "scrollmouse"].includes(name)) return await scrollMouse(params);
    if (["type", "typetext", "text"].includes(name)) return await typeText(params);
    if (["keypress", "presskey", "key", "sendkey"].includes(name)) return await keypress(params);
    if (["close", "shutdown"].includes(name)) return { ok: true };
    await log({ event: "unknown-method", method, params });
    throw new Error(`Unsupported local Computer Use helper method: ${method}`);
  }

  async close() {
    await log({ event: "transport-closed" });
  }
}
